content/blog/valabot.md

```diff
@@ -0,0 +1,68 @@
+
+I fine-tuned the model on Vala programming language datasets. 
+This involved downloading as many Vala projects as I could find from GitHub, extracting the Vala source files and splitting them into ~40 line segments. 
+I then used Llama3 to create logical and predictable "holes" in each segment, which were then used to create the FIM (fill-in-the-middle) dataset. 
+This data preparation process took 96 hours of GPU time using my quad-RX6800 machine over a weekend. 
+The resulting dataset was cleaned to remove non-code elements, such as license headers, and personal identifiable information like names and email addresses, and some old auth secrets that should never have been published! 
+
+# The Training
+
+The fine-tuning process took 10 hours on an RTX 3090. The result was a LoRA, which was merged back into the base model, converted to GGUF and quantised to q8_0 which is the format required by TabbyML. 
+
+# The Result
+
+The outcome was a model that is more helpful and productive for Vala-related projects. By fine-tuning Deepseek Coder, I was able to create a more accurate and effective AI coding assistant that understands the nuances of the Vala programming language.
+
+The model is hosted on Huggingface. <https://huggingface.co/scowen/deepseek-coder-6.7b-vala/tree/main>
+
+It can be used in VSCode, VIM and other popular IDEs with TabbyML. The complete instructions, training scripts and dataset are available here: <https://github.com/supercamel/ValaBot>
+
+# Licensing
+
+There has been some discussion about the issue of licensing and fair use of GPL code in particular to fine tune a language model. The training data is quite obviously derived from projects under various licenses, the most restrictive of which is the GPL. As a derived work, the training data must also be GPL'd. One might even argue that the LoRA is a derived work and must be GPL'd. None of this should be particularly concerning to anybody. 
+
+As far as the code produced by the model goes, I believe it is owned by the user of the model. Whoever provides the prompts owns the output and may license it as they wish. I base this belief on the fact that while the GPL doesn't explicitly address AI, the FSF does have this to say
+```
+ The output of a program is not, in general, covered by the copyright on the code of the program. So the license of the code of the program does not apply to the output ...
+```
+Source: <https://www.gnu.org/licenses/gpl-faq.en.html#WhatCaseIsOutputGPL>
+
+While the model has been trained on GPL'd code, I consider the code that is produced by the model to be an output. It simply won't produce a full screen of code from a GPL'd project. The training data essentially just biases the model to produce code that is a bit more like Vala than the Java and C# that it is trained on - it won't accurately reproduce the GPL'd code it was fine-tuned on in any meaningful quantity. 
```

Comment:
I don't think I can personally approve this pull request. Adding to what the blog post said, licensing in language models is a controversial topic in the FOSS community.

I would've been fine with it if the model wasn't trained on GPL'd code.

Better to post this in a personal blog post in my opinion

Reply:
Is there a place where this topic is being discussed by the FOSS community? I'm interested in reading up more and hearing the positions of people in the FOSS community. 

In any case, AI coding assistants will not be going away. It's happening on a massive scale commercially and many people are paying for the convenience of AI trained on GPL'd code. If it's controversial that someone would optimize a free, open weight model for Vala for the benefit of all who might use it with the intent to *replace* a commercial option, then I suggest we're loading a foot-gun.

Reply:
> Is there a place where this topic is being discussed by the FOSS community? I'm interested in reading up more and hearing the positions of people in the FOSS community.

Sadly, I haven't been able to provide blog posts or bookmark threads on this. Main concerns have been with the licenses of code being used for training and it's transparency however a search is the best I can do now for you to get an idea of the discourse: https://mastodon.social/explore -> "LLM license' and "LLM License'. 

>  The situation today is that many people are paying for the convenience of AI trained on GPL'd code. If it's controversial that someone would optimize a free, open weight model for Vala for the benefit of all who might use it with the intent to replace a commercial option, then I suggest we're loading a foot-gun.

Yes, this is why I liked the idea of this blog post being created in the first place :)

> All the repositories used to do the fine tuning are listed. I've done this with the maximum amount of transparency and my intent is completely inline with GPL & FOSS values.

> Now that I read it back, the licensing section does come across as defensive and doesn't exactly say what it should. Perhaps I should rewrite the licensing section to improve the messaging?

That's a great idea!

Reply:
I've rewritten it, and put a little more effort into the licensing section. What do you think?